File: docs/integration_toolbox/devpilot_external_client.js
import crypto from "node:crypto";

function cleanBaseUrl(value) {
  const baseUrl = String(value || "").trim().replace(/\/+$/, "");
  if (!baseUrl) throw new Error("DEVPILOT_API_BASE_URL is required");
  return baseUrl;
}

function requireValue(name, value) {
  const text = String(value || "").trim();
  if (!text) throw new Error(`${name} is required`);
  return text;
}

function jsonHeaders(settings, idempotencyKey) {
  return {
    "Content-Type": "application/json",
    "X-DevPilot-Source-System": requireValue("DEVPILOT_SOURCE_SYSTEM", settings.sourceSystem),
    "X-DevPilot-Api-Key": requireValue("DEVPILOT_API_KEY", settings.apiKey),
    "X-DevPilot-Request-Id": crypto.randomUUID(),
    "X-DevPilot-Idempotency-Key": idempotencyKey
  };
}

async function requestJson(url, options = {}) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), options.timeoutMs || 10000);
  try {
    const response = await fetch(url, { ...options, signal: controller.signal });
    const payload = await response.json().catch(() => ({}));
    if (!response.ok) {
      throw new Error(`DevPilot request failed: ${response.status} ${payload.error || "unknown_error"}`);
    }
    return payload;
  } finally {
    clearTimeout(timeout);
  }
}

export async function testDevPilotConnection(settings) {
  const baseUrl = cleanBaseUrl(settings.baseUrl);
  return requestJson(`${baseUrl}/api/external/projects`, {
    method: "GET",
    headers: {
      "X-DevPilot-Source-System": requireValue("DEVPILOT_SOURCE_SYSTEM", settings.sourceSystem),
      "X-DevPilot-Api-Key": requireValue("DEVPILOT_API_KEY", settings.apiKey)
    }
  });
}

export async function registerDevPilotProject(settings, project) {
  const baseUrl = cleanBaseUrl(settings.baseUrl);
  const sourceSystem = requireValue("DEVPILOT_SOURCE_SYSTEM", settings.sourceSystem);
  const externalProjectId = requireValue("external_project_id", project.external_project_id);
  return requestJson(`${baseUrl}/api/external/projects/register`, {
    method: "POST",
    headers: jsonHeaders(settings, `register:${sourceSystem}:${externalProjectId}`),
    body: JSON.stringify(project)
  });
}

export async function sendDevPilotProjectEvent(settings, externalProjectId, event) {
  const baseUrl = cleanBaseUrl(settings.baseUrl);
  const sourceSystem = requireValue("DEVPILOT_SOURCE_SYSTEM", settings.sourceSystem);
  const projectId = encodeURIComponent(requireValue("external_project_id", externalProjectId));
  const eventType = requireValue("event_type", event.event_type);
  const keySeed = event.idempotencyKey || `${eventType}:${event.commit_sha || event.message || crypto.randomUUID()}`;
  return requestJson(`${baseUrl}/api/external/projects/${projectId}/events`, {
    method: "POST",
    headers: jsonHeaders(settings, `event:${sourceSystem}:${externalProjectId}:${keySeed}`),
    body: JSON.stringify(event)
  });
}

export async function createDevPilotHandoff(settings, taskId, handoff) {
  const baseUrl = cleanBaseUrl(settings.baseUrl);
  const sourceSystem = requireValue("DEVPILOT_SOURCE_SYSTEM", settings.sourceSystem);
  const idempotencySeed = handoff.external_ref || handoff.reason || crypto.randomUUID();
  return requestJson(`${baseUrl}/api/external/tasks/${encodeURIComponent(taskId)}/handoffs`, {
    method: "POST",
    headers: jsonHeaders(settings, `handoff:${sourceSystem}:${taskId}:${idempotencySeed}`),
    body: JSON.stringify(handoff)
  });
}

export async function generateDevPilotAiText(settings, request) {
  const baseUrl = cleanBaseUrl(settings.baseUrl);
  const sourceSystem = requireValue("DEVPILOT_SOURCE_SYSTEM", settings.sourceSystem);
  const externalRef = request.external_ref || request.externalRef || crypto.randomUUID();
  const body = {
    provider: request.provider || "openai",
    model: request.model || "gpt-4.1-mini",
    capability: request.capability || "generate",
    prompt: requireValue("prompt", request.prompt),
    external_ref: externalRef,
    metadata: request.metadata || {}
  };
  return requestJson(`${baseUrl}/api/external/ai/generate`, {
    method: "POST",
    headers: jsonHeaders(settings, `ai-generate:${sourceSystem}:${externalRef}`),
    body: JSON.stringify(body)
  });
}

// Never log settings.apiKey or full request headers.

File: docs/integration_toolbox/devpilot_external_client.py
import os
import uuid

import requests


class DevPilotClient:
    def __init__(self, base_url=None, source_system=None, api_key=None, timeout=10):
        self.base_url = (base_url or os.getenv("DEVPILOT_API_BASE_URL", "")).rstrip("/")
        self.source_system = (source_system or os.getenv("DEVPILOT_SOURCE_SYSTEM", "")).strip()
        self.api_key = (api_key or os.getenv("DEVPILOT_API_KEY", "")).strip()
        self.timeout = timeout
        if not self.base_url:
            raise ValueError("DEVPILOT_API_BASE_URL is required")
        if not self.source_system:
            raise ValueError("DEVPILOT_SOURCE_SYSTEM is required")
        if not self.api_key:
            raise ValueError("DEVPILOT_API_KEY is required")

    def _headers(self, idempotency_key=None, json_body=False):
        headers = {
            "X-DevPilot-Source-System": self.source_system,
            "X-DevPilot-Api-Key": self.api_key,
            "X-DevPilot-Request-Id": str(uuid.uuid4()),
        }
        if idempotency_key:
            headers["X-DevPilot-Idempotency-Key"] = idempotency_key
        if json_body:
            headers["Content-Type"] = "application/json"
        return headers

    def _request(self, method, path, **kwargs):
        response = requests.request(
            method,
            f"{self.base_url}{path}",
            timeout=self.timeout,
            **kwargs,
        )
        try:
            payload = response.json()
        except ValueError:
            payload = {}
        if response.status_code >= 400:
            error = payload.get("error", "unknown_error")
            raise RuntimeError(f"DevPilot request failed: {response.status_code} {error}")
        return payload

    def test_connection(self):
        return self._request("GET", "/api/external/projects", headers=self._headers())

    def register_project(self, project):
        external_project_id = str(project.get("external_project_id") or "").strip()
        if not external_project_id:
            raise ValueError("external_project_id is required")
        idempotency_key = f"register:{self.source_system}:{external_project_id}"
        return self._request(
            "POST",
            "/api/external/projects/register",
            headers=self._headers(idempotency_key=idempotency_key, json_body=True),
            json=project,
        )

    def send_project_event(self, external_project_id, event):
        external_project_id = str(external_project_id or "").strip()
        event_type = str(event.get("event_type") or "").strip()
        if not external_project_id:
            raise ValueError("external_project_id is required")
        if not event_type:
            raise ValueError("event_type is required")
        seed = event.get("idempotency_key") or event.get("commit_sha") or event.get("message") or str(uuid.uuid4())
        idempotency_key = f"event:{self.source_system}:{external_project_id}:{event_type}:{seed}"
        return self._request(
            "POST",
            f"/api/external/projects/{external_project_id}/events",
            headers=self._headers(idempotency_key=idempotency_key, json_body=True),
            json=event,
        )

    def create_handoff(self, task_id, handoff):
        task_id = str(task_id or "").strip()
        if not task_id:
            raise ValueError("task_id is required")
        seed = handoff.get("external_ref") or handoff.get("reason") or str(uuid.uuid4())
        idempotency_key = f"handoff:{self.source_system}:{task_id}:{seed}"
        return self._request(
            "POST",
            f"/api/external/tasks/{task_id}/handoffs",
            headers=self._headers(idempotency_key=idempotency_key, json_body=True),
            json=handoff,
        )

    def generate_ai_text(self, prompt, provider="openai", model="gpt-4.1-mini", capability="generate", external_ref=None, metadata=None):
        prompt = str(prompt or "")
        if not prompt.strip():
            raise ValueError("prompt is required")
        stable_ref = external_ref or str(uuid.uuid4())
        idempotency_key = f"ai-generate:{self.source_system}:{stable_ref}"
        return self._request(
            "POST",
            "/api/external/ai/generate",
            headers=self._headers(idempotency_key=idempotency_key, json_body=True),
            json={
                "provider": provider,
                "model": model,
                "capability": capability,
                "prompt": prompt,
                "external_ref": stable_ref,
                "metadata": metadata or {},
            },
        )


# Never log client.api_key or full request headers.
